pkg/ddl/backfilling_dist_scheduler.go

```diff
@@ -466,32 +466,42 @@ func generateMergePlan(
 ) ([][]byte, error) {
 	// check data files overlaps,
 	// if data files overlaps too much, we need a merge step.
-	multiStats := make([]external.MultipleFilesStat, 0, 100)
+	var multiStatsGroup [][]external.MultipleFilesStat
 	var kvMetaGroups []*external.SortedKVMeta
 	err := forEachBackfillSubtaskMeta(taskHandle, task.ID, proto.BackfillStepReadIndex,
 		func(subtask *BackfillSubTaskMeta) {
 			if kvMetaGroups == nil {
 				kvMetaGroups = make([]*external.SortedKVMeta, len(subtask.MetaGroups))
+				multiStatsGroup = make([][]external.MultipleFilesStat, len(subtask.MetaGroups))
 			}
 			for i, g := range subtask.MetaGroups {
 				if kvMetaGroups[i] == nil {
 					kvMetaGroups[i] = &external.SortedKVMeta{}
+					multiStatsGroup[i] = make([]external.MultipleFilesStat, 0, 100)
 				}
 				kvMetaGroups[i].Merge(g)
-				multiStats = append(multiStats, g.MultipleFilesStats...)
+				multiStatsGroup[i] = append(multiStatsGroup[i], g.MultipleFilesStats...)
 			}
 		})
 	if err != nil {
 		return nil, err
 	}
-	if skipMergeSort(multiStats) {
+
+	allSkip := true
+	for _, multiStats := range multiStatsGroup {
+		if !skipMergeSort(multiStats) {
+			allSkip = false
+			break
+		}
+	}
+	if allSkip {
 		logger.Info("skip merge sort")
 		return nil, nil
 	}
 
-	// generate merge sort plan.
-	dataFiles := make([]string, 0, 1000)
+	metaArr := make([][]byte, 0, 16)
 	for i, g := range kvMetaGroups {
+		dataFiles := make([]string, 0, 1000)
```

Comment:
Why not generate subtasks for each index that really needs merge sort？

Reply:
If we do this, in step 3, we need to scan subtasks meta from both step 2 and step 1.